do not lower the case of extension id

sandy081 opened the pull request:
This PR fixes #140120

Extension id is constructed as it is from the `package.json` while running extensions in desktop. But in web  extension id is lower cased, hence this caused the reported issue. 

Fix is to construct the extension id in web in the same way as in desktop. But this causes data loss for the extension in web. Becuase the extension memento and storage are based on extension id and changing the cases will point the memento and storage to differen keys/locations.

Hence added migration step in web and it is done only if constructed extension id is not lower cased. Migration includes

- Moving global and workspace memento and storage from old key to new key

This migration step happens only once as it is remembered at global and workspace level

bpasero commented:
I fail to understand how the ID of extension memento is any relevant unless we have 2 extensions with the same identifier but different casing competing? For an extension the way how we store the memento should be irrelevant because all they get is an object to set properties to, no?

Adding a second pair of eyes for review. And I think this needs some kind of integration test if possible.

bpasero commented on src/vs/workbench/api/browser/mainThreadStorage.ts:
```diff
@@ -67,7 +76,62 @@ export class MainThreadStorage implements MainThreadStorageShape {
 		this._extensionsStorageSyncService.setKeysForSync(extension, keys);
 	}
+
+	private async _migrateExtensionStorage(from: string, to: string, storageMigratedKey: string): Promise<void> {
```
This encodes a lot of knowledge about how storage works in web. For example on desktop, we could never just rename a SQLite DB from one to the other name without properly closing the DB and ensuring it is not locked. I almost believe that migration needs to be implemented within the storage service itself and not via file services from the outside. This would then throw an error from the desktop based implementation.

bpasero replied:
Instead of moving files on disk, can we not recreate the storage value with the right name and the right properties and let the storage service itself deal with the file management?

sandy081 replied:
Not sure what you mean by assuming storage implementation here. If you see, I am storage service APIs to update the mementos and storage URIs (globalStorage & worksapcestorage) are file based both in desktop and web. This migration is triggered only in web because, the desktop does not have this issue and I do not want to have unnecessary overload in that case

jrieken commented:
> But this causes data loss for the extension in web. 

How often are upper-case characters used in extension ids? Knowing that and that given that web is relatively fresh might allow us loss data.

sandy081 commented:
> I fail to understand how the ID of extension memento is any relevant unless we have 2 extensions with the same identifier but different casing competing?

The actual fix for #140120 can cause the same extension to have different mementos before and after the fix. To explain in detail, in web, the extension id of `GitHub.remotehub` extension is lower cased which is `github.remotehub`. So currently the memento state of this extension is stored using the key `github.remotehub`.  With the fix, the extension id is not lower cased and remains as is which is `GitHub.remotehub`. So when the extension runs with the fix, the memento state is fetched using the key `GitHub.remotehub` which is not same as before. Hence the extension looses its data. This is same with `storageUri` which has files stored by the extension per workspace. 

The fix without migration will cause data loss for `GitHub.remotehub` extension as it stored unresolved files and changes in memento and workspace storage. Actually I pushed this fix sometime back and reverted it as soon as we saw this data loss.

> How often are upper-case characters used in extension ids? Knowing that and that given that web is relatively fresh might allow us loss data.

I do not have numbers but definitely all GitHub extensions as upper case publisher id and there are others too. I think data loss is severe here because as I said before GitHub.remotehub extension stores changes here and loosing it might be quite annoying to user.

bpasero commented:
@sandy081 I am still puzzled. 

Let me try to recap what your code is doing:
* using `IStorageService` you migrate the key `<extension ID lowercase>` to `<extension ID realcase>` in global and workspace storage (this maps to `ExtensionContext.globalState` and `ExtensionContext.workspaceState`)
* using `IFileService` you migrate the `globalStorage/<extension ID lowercase>` and `workspaceStorage/<workspaceId>/<extension ID lowercase>` to `<extension ID realcase>` (this maps to `ExtensionContext.globalStorageUri` and `ExtensionContext.workspaceStorageUri`)

At first I was thinking this is only a storage service migration, but in fact this is 2 migrations of 2 entirely different things in the same method. This was hard to understand for me at first because `mainThreadStorage` is about mementos and not about the `globalStorageUri` and `workspaceStorageUri`.

So why is the migration of `globalStorageUri` and `workspaceStorageUri` triggered from `$getValue`? I think that makes no sense because those properties are used in other contexts. I think if we want to migrate that, we need to do that before any extension is being activated. 

**Storage Service Migration**
Why do we need to do this though? For an extension the way how mementos are persisted is not relevant. The fact that we store them in storage service is an implementation detail. We can easily continue to use the lowercase form and not migrate. I.e. we could just lowercase the extension ID in `mainThreadStorage` to preserve the data without the need for a migration.

**Storage URI Migration**
This is a path on disk that extensions can use to persist any amount of data. It is backed by a file system on desktop (local disk) but IndexDB (user data) on web. Same idea here: could we just keep the extension ID lowercase, given extensions just store stuff in those locations? Do they care about the casing of the parent folder they persist to?

And if we must migrate this data I think we have to do it before any extension is being activated.

sandy081 commented:
@bpasero made changes as we discussed today.

- Extracted extension storage related methods into `IExtensionStorageService`. I was having `IExtensionsStorageSyncService` for Settings Sync purpose. I merged this service into `IExtensionStorageService` and removed some redundant util functions in settings sync land in favour of this new service.

- Moved extension storage migration into a util file which can be reused for other migration scenarios I have (migrating nightly/insiders to pre-release extensions)

- renamed `$getValue` method to `initializeExtensionStorage` and `key` to `extensionid`

Please take a look.

bpasero reviewed (approved):
Only glanced over the changes high level, didn't do a deep line by line review because I trust you know what you are doing and the previous review looked fine to me, at least for the migration code.

Would it at all be possible to write an integration tests for this, maybe by renaming the storage location to lower case before starting the test?

I would add a TODO@sandy to remember to remove this code in 6 months or have an issue for it.

bpasero commented:
Btw it looks like the call to `migrateStorage` is not awaited, there are 2 locations beginning from https://github.com/microsoft/vscode/pull/140122/files#diff-7b81baaeef780c48f07d57c4e97e78d33aed5a2b5ec97e8c7996e49facc90c1dR67

bpasero reviewed (changes requested):
Actually my last comment seems important to fix.

sandy081 commented:
Good catch, missed it when I extracted the common code in to a function. 

I will write unit tests `migrateExtensionStorage` util function

sandy081 commented:
@bpasero updated. Please take a look.
